src/gluonts/distribution/distribution.py

```diff
@@ -83,26 +93,42 @@ def crps(self, x: Tensor) -> Tensor:
         """
         raise NotImplementedError()
 
-    def loss(self, x: Tensor) -> Tensor:
+    def _loss(self, x: Tensor) -> Tensor:
+        return -self.log_prob(x)
+
+    def loss(self, x: Tensor, mask: Tensor = None) -> Tensor:
         r"""
         Compute the loss at `x` according to the distribution.
 
         By default, this method returns the negative of `log_prob`. For some
         distributions, however, the log-density is not easily computable
         and therefore other loss functions are computed.
 
+        Additionally an optional Mask can be specified to mask the loss.
+
         Parameters
         ----------
-        x
+        x:
             Tensor of shape `(*batch_shape, *event_shape)`.
+        mask:
+            Tensor of shape `(*batch_shape, *event_shape)` that defines the mask.
+            Values in the loss corresponding to a `0` in the mask will be set to `0`,
+            other values are left unchanged.
 
         Returns
         -------
         Tensor
             Tensor of shape `batch_shape` containing the value of the loss
             for each event in `x`.
         """
-        return -self.log_prob(x)
+        F = getF(x)
+        if mask is not None:
+            x = F.where(condition=mask, x=x, y=self.point_in_support)
+            loss = self._loss(x)
+            loss = F.where(condition=mask, x=loss, y=F.zeros_like(x))
```

Comment:
Why is `point_in_support` necessary here? Shouldn't

```
F.where(condition=mask, x=loss, y=F.zeros_like(x))
```
be sufficient? Can you maybe give a minimal example with some nd arrays?

Reply:
TLDR: essentially the problem is that we get `nan`s upon back-propagation if we only use `F.where`:

If we have `0` due to padding in the data and we calculate the `-log_loss(x)` for the Beta or Gamma Distribution we get `-inf`, which we want to mask.

So in the forward pass everything works out with `F.where` and we get valid loss values, however, upon back-propagation we get `-inf * 0 = nan` because of the way the gradient is calculated: 

<img width="128" alt="Screenshot 2019-12-17 at 21 49 48" src="https://user-images.githubusercontent.com/25365592/71253484-5500c800-2328-11ea-861c-407325ffd1cd.png">

Because `F.where` is basically just another network layer and because `-inf * 0` yields `nan` and not `0`.

Apparently this is a known issue: 
https://github.com/pytorch/pytorch/issues/23156

Reply:
@vafl there are issues if some elements of `x` are imputed with values that are not in the support of the distribution. For example, in `Beta` or `Gamma` zero is not in the support, but `nan`s in the data are imputed with zeros. This is fine in the forward pass: you'll get some `inf` in the log-likelihood which will be masked in this line, as you say. But this creates problems in the backward pass, apparently, which can only be avoided by not producing the `inf` in the first place. Aaron found a similar issue in pytorch: https://github.com/pytorch/pytorch/issues/23156

A MWE for this issue is the following:

```python
import mxnet as mx
from mxnet import autograd

param = mx.nd.array([1.0])
y = mx.nd.array([0.0])

param.attach_grad()

with autograd.record():
    loss = mx.nd.log(y*param)
    mask = mx.nd.array([0.0])
    masked_loss = mx.nd.where(condition=mask, x=loss, y=mx.nd.zeros_like(loss))

print(masked_loss)

dparam = autograd.grad(masked_loss, [param])[0]

print(dparam)
```

Reply:
The above prints out

```
[0.]
<NDArray 1 @cpu(0)>

[nan]
<NDArray 1 @cpu(0)>
```

Reply:
I see, thanks for clarifying this! One alternative design would be to allow passing `nan`'s in the input and then do all the masking and logic of selecting a dummy value inside `log_prob`.

Reply:
@vafl I think the problem with that is that nans need to be out of the way once the data reaches the network, or did I misunderstand?

Reply:
AFAIK you can pass nans in via the data. I think we used this at some point. I'm not sure whether it is a common practice, but in the end nans should be just floats.

Reply:
But if a nan goes as input to any block, it will likely result in nans coming out of it. Or do you mean substitute nans with zeros in the input (past) but leave nans in the label (future)?

Reply:
Ok, yeah. Handling nans inside the graph may be tricky with this gradient issue.